Add Vivid image

cb372 opened the pull request:
Adds a builder that creates an Ubuntu 15.04 (Vivid) image with Java 8 installed.

TODO:
- [x] systemd support for native-packager feature
- [x] manually test that all features work
- [x] uncomment the `ixgbevf` stuff

Unfortunately Ubuntu 15.04 also has an old version of `ixgbevf` so we will need to keep the code that upgrades it. But that code is currently commented out because SourceForge is down :(

Commit:
Add a build for an Ubuntu 15.04 (Vivid) image With Java 8

Commit:
Pass command line args to Packer

This is useful when running locally, e.g. you can run

```
bash build.sh default -only=ubuntu-vivid-java8
```

Commit:
Add systemd service file for Logstash feature

cb372 commented on packer/resources/features/logstash/install.sh:
```diff
@@ -18,9 +18,22 @@ ln -s /opt/features/logstash/logstash /opt/logstash
+sed \
+  -e "s/@USER@/$ESCAPED_USER/g" \
+  -e "s/@CONFIGFILE@/$ESCAPED_CONFIGFILE/g" \
+  /opt/features/logstash/logstash.service.template > /etc/systemd/system/logstash.service
```
Can @sihil or @philwills confirm that `/etc/systemd/system` is a reasonable place to put the service file?

philwills replied:
It works there. sbt-native-packager goes for /usr/lib/systemd/system/, but I don't have a strong opinion either way.

philwills commented on packer/resources/features/logstash/install.sh:
```diff
@@ -18,9 +18,22 @@ ln -s /opt/features/logstash/logstash /opt/logstash
 ESCAPED_CONFIGFILE=$(echo $CONFIGFILE | sed -e 's/[\/&]/\\&/g')
+
+# Upstart
+if [ -d "/etc/init" ]; then
```
There is still an `/etc/init` directory on vivid as far as I can see. Does it make sense to do both?

cb372 replied:
I'm trying to make the features work on both the Trusty and Vivid images. But you're right, it might be pretty confusing to have the service installed in 2 different ways on Vivid.

I'll change the logic to "if there's a systemd directory, install a systemd script, otherwise install an upstart script".

philwills commented on packer/resources/ubuntu-vivid.sh:
```diff
@@ -0,0 +1,86 @@
+apt-get --yes --force-yes install \
+  git wget language-pack-en build-essential python-setuptools \
+  openjdk-8-jre-headless openjdk-8-jdk cloud-guest-utils jq \
+  ntp unzip python3-pip
```
I don't think ntp is needed anymore, as it or equivalent already installed by Vivid, but probably no harm.

What's git for?

cb372 replied:
> What's git for?

dunno, but it might be handy

philwills replied:
Feels like an odd thing to have on a production box to me, but I may be in the minority on that one.

sihil replied:
IIRC git is there because some installation processes use git to install. Bower and friends spring to mind?

To be honest though, I don't recall any specific example.

philwills replied:
We're not running bower on the machine for anything are we?

I'd certainly hope we were baking those dependencies in at artifact build time.

sihil replied:
I think a few hacks do this to bootstrap boxes. I don't think any prod servers do. Happy to take it out and see what breaks.

sihil replied:
@philwills NTP isn't installed by default on Vivid - can you elaborate on your statement, I don't particularly want two time sync services fighting over the clock.

philwills replied:
https://lists.ubuntu.com/archives/ubuntu-devel/2014-December/038569.html was what I'd found, but I might be wrong about it actually getting in.

sihil replied:
Googling doesn't bring anything up about this and it certainly isn't a default. Would have expected it to have been shouted from the rooftops as the suggestion on the mailing list is a pretty big change. Going to assume it's not there and carry on as I was.

philwills replied:
Yea, probably just me getting overexcited about someone proposing a sane default.

sihil replied:
It's an interesting conversation - I suspect it failed to get in as people didn't think NTP was a secure enough service to install by default. Looks like there is a component of systemd that does time syncing as well - presumably using the NTP protocol regardless, which is deemed pretty much irreparably insecure from what I understand.

sihil committed:
Use an env var instead of parameters

The first parameter is optional, which makes it impossible to sanely
deal with flags in this way

sihil committed:
Merge branch 'master' into add-vivid-image

sihil committed:
Update systemd detection to use /proc/1/comm, add native-packager support

sihil commented:
@cb372 What else needs to be done for the `ixgbevf` module? Simply uncomment?

sihil commented:
I've checked the the gmn-dns feature works fine. The ELK stuff won't work on Vivid - I suggest that we merge without that and ask @andybotting to investigate and migrate over.

cb372 commented:
@sihil thanks for taking over this branch.

Yes, SourceForge (sigh) is back up now, so just uncomment. As far as I know the same driver and patch should work on Vivid, but I haven't been able to test it yet.

I'd be tempted to just stick the driver tarball into this repo, rather than depending on SourceForge.

andybotting commented:
@shil I'll start working on ELK stack. Btw, I noticed that @satterly has done a lot of work on his own ELK stack, also running Kibana 4

sihil committed:
Include systemd template, fix install.sh ordering

sihil committed:
Re-introduce enhanced networking

sihil commented:
Thanks @andybotting. BTW it would be great to split the ES cluster from the frontend stuff. As discussed it would make it far quicker to roll out new versions of Kibana etc.

I think @satterly doesn't need the encryption which makes it specific to the guardian (or relies on the machine image repo - the images from which are not public).

sihil commented:
Features:
- ebs - will leave to @andybotting 
- elk-stack - will leave to @andybotting 
- gnm-ca - confirmed working
- gnm-dns - confirmed working
- logstash - changed, but will leave to @cb372 
- native-packager - confirmed working
- rewrite-apt-sources - confirmed working
- templating - used by native-packager / ebs (thus worked based on native packager working)

sihil committed:
Re-comment line that was acutally a comment

cb372 commented:
Looks good, but please don't merge yet. I'm halfway through checking that Logstash works as expected with systemd. I'll finish tomorrow.

Commit:
Tweak Logstash systemd file

cb372 commented:
OK, all sorted. Merge when ready.

mchv commented on packer/resources/ubuntu-vivid.sh:
```diff
@@ -0,0 +1,85 @@
+wget http://sourceforge.net/projects/e1000/files/ixgbevf%20stable/2.16.1/ixgbevf-2.16.1.tar.gz
+tar -xzf ixgbevf-2.16.1.tar.gz
+pushd ./ixgbevf-2.16.1/src
+wget "https://gist.githubusercontent.com/defila-aws/44946d3a3c0874fe3d17/raw/af64c3c589811a0d214059d1e4fd220a96eaebb3/patch-ubuntu_14.04.1-ixgbevf-2.16.1-kcompat.h.patch" -O patch.kcompat.h
```
Sorry to comment late but I think this patch is not necessary any more as the kernel version provided in `Vivid` is `3.19` and the patch was specifically for kernel below `3.15`

sihil replied:
Any idea how I can confirm that? Am doing some more work on this in #37

mchv replied:
The driver code was [not compiling without the patch](https://bugs.launchpad.net/ubuntu-on-ec2/+bug/1254930/comments/10), so if it is compiling it should be fine.
